test/extensions/tracers/datadog/event_scheduler_test.cc

```diff
@@ -0,0 +1,111 @@
+#include <chrono>
+
+#include "source/extensions/tracers/datadog/event_scheduler.h"
+
+#include "test/mocks/event/mocks.h"
+#include "test/mocks/thread_local/mocks.h"
+
+#include "datadog/json.hpp"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace Datadog {
+namespace {
+
+TEST(DatadogEventSchedulerTest, ScheduleRecurringEventCallsCreatesATimer) {
+  testing::NiceMock<ThreadLocal::MockInstance> thread_local_storage_;
+
+  EventScheduler scheduler{thread_local_storage_.dispatcher_};
+  testing::MockFunction<void()> callback;
+  // The interval is arbitrary in these tests; we just have to be able to
+  // compare it to what was passed to the mocks.
+  // The only requirement is that it be divisible by milliseconds, because
+  // that's what `Timer::enableTimer` accepts.
+  const auto interval = std::chrono::milliseconds(2000);
+
+  EXPECT_CALL(thread_local_storage_.dispatcher_, createTimer_(testing::_));
+
+  (void)scheduler.schedule_recurring_event(interval, callback.AsStdFunction());
+}
+
+// This could be tested above, but introducing an `Event::MockTimer` disrupts
+// our ability to track calls to `MockDispatcher::createTimer_`. So, two
+// separate tests.
+TEST(DatadogEventSchedulerTest, ScheduleRecurringEventEnablesATimer) {
+  testing::NiceMock<ThreadLocal::MockInstance> thread_local_storage_;
+  auto* const timer_ = new testing::NiceMock<Event::MockTimer>(&thread_local_storage_.dispatcher_);
```

Comment:
nit: don't use an underscore suffix for a local. Comment on memory management: who deletes timer?

Reply:
This pattern is copied from here: https://github.com/envoyproxy/envoy/blob/40d15a9c391c91b05353abdbef78407ff49972fa/test/extensions/tracers/datadog/datadog_tracer_impl_test.cc#L54

Which was copied from here: https://github.com/envoyproxy/envoy/blob/40d15a9c391c91b05353abdbef78407ff49972fa/test/extensions/tracers/zipkin/zipkin_tracer_impl_test.cc#L55

It occurs to me now that the raw pointer is probably a substitute for "optional" in the original code.

> who deletes timer?

Nobody, it's leaked, as in the copied-from code.

I'll see if it works the same with just a `testing::NiceMock<Event::MockTimer>` on the stack (no pointer). That'd be better.

Reply:
Ah, it's not for lack of optional. This is indeed the intended usage of `MockTimer`: https://github.com/envoyproxy/envoy/blob/eee9b7cf37545147e135a81624589de7465179e9/test/mocks/event/mocks.h#L183-L186